### PR TITLE
[libc][stdfix] Add integer square root with fixed point output functions.

### LINKED ISSUE #83924: Support integral square root with unsigned fixed point output in llvm-libc
`sqrtuk` and co work really well but we don't have `sqrtulk` at the moment, so none of the fixed point sqrt functions will work out of the box for values larger than 2^16. Even though it's not part of ISO 18037, it would be nice to have an integral sqrt as an option.

### PR DESCRIPTION
**Pull request by lntue**
Fix https://github.com/llvm/llvm-project/issues/83924.

### CODE REVIEW

**Comment by nickdesaulniers** [comment]
> Even though it's not part of ISO 18037

What does that mean? Are these now functions outside of that spec? :unamused:

**Comment by lntue** [comment]
> > Even though it's not part of ISO 18037
> 
> What does that mean? Are these now functions outside of that spec? 😒

Technically, the ISO standard does not define any math functions beyond mixed integer-and-fixed-point multiplications and divisions, not even including `sqrt`.

Ideally, I would like to have all C23 math functions available for all numeric types.  We could have a complete math library for fixed point type and with the proof of their usefulness, maybe fixed point types + its math library will be incorporated into the future C standards, and ours can be served as a reference implementation?

**Comment by nickdesaulniers** [comment]
> > > Even though it's not part of ISO 18037
> > 
> > 
> > What does that mean? Are these now functions outside of that spec? 😒
> 
> Technically, the ISO standard does not define any math functions beyond mixed integer-and-fixed-point multiplications and divisions, not even including `sqrt`.
> 
> Ideally, I would like to have all C23 math functions available for all numeric types. We could have a complete math library for fixed point type and with the proof of their usefulness, maybe fixed point types + its math library will be incorporated into the future C standards, and ours can be served as a reference implementation?

Can we make that more explicit on https://libc.llvm.org/math/stdfix.html?

File: libc/spec/stdc_ext.td [hunk 1]
```diff
@@ -55,6 +55,9 @@ def StdcExt : StandardSpec<"stdc_ext"> {
           GuardedFunctionSpec<"sqrtuhk", RetValSpec<UnsignedShortAccumType>, [ArgSpec<UnsignedShortAccumType>], "LIBC_COMPILER_HAS_FIXED_POINT">,
           GuardedFunctionSpec<"sqrtuk", RetValSpec<UnsignedAccumType>, [ArgSpec<UnsignedAccumType>], "LIBC_COMPILER_HAS_FIXED_POINT">,
           GuardedFunctionSpec<"sqrtulk", RetValSpec<UnsignedLongAccumType>, [ArgSpec<UnsignedLongAccumType>], "LIBC_COMPILER_HAS_FIXED_POINT">,
+
+          GuardedFunctionSpec<"uhksqrtus", RetValSpec<UnsignedShortAccumType>, [ArgSpec<UnsignedShortType>], "LIBC_COMPILER_HAS_FIXED_POINT">,
```
Comment (michaelrj-google) [inline libc/spec/stdc_ext.td:59]: is this actually standardized or should this go in `llvm_libc_ext.td`?

File: libc/test/src/stdfix/ISqrtTest.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  using OutType =
+      typename LIBC_NAMESPACE::fixed_point::internal::SqrtConfig<T>::OutType;
+  using FXRep = LIBC_NAMESPACE::fixed_point::FXRep<OutType>;
+  static constexpr OutType zero = FXRep::ZERO();
+  static constexpr OutType one = static_cast<OutType>(1);
+  static constexpr OutType eps = FXRep::EPS();
+
+public:
+  typedef OutType (*SqrtFunc)(T);
```
Comment (michaelrj-google) [inline libc/test/src/stdfix/ISqrtTest.h:28]: do these tests properly handle the fast math case?

**Review by michaelrj-google — APPROVED** [review]
Overall LGTM with some nits

File: libc/spec/stdc_ext.td (hunk 1 above)
Comment (nickdesaulniers) [reply libc/spec/stdc_ext.td:59]: right, I would prefer stronger language / guarantees between "implemented against a spec" vs straight up "extension."

both in documentation and in our headergen files as @michaelrj-google suggested.

Comment (lntue) [reply libc/spec/stdc_ext.td]: I've moved the specs of these functions to `llvm_libc_ext.td`.

**Comment by lntue** [comment]
> > > > Even though it's not part of ISO 18037
> > > 
> > > 
> > > What does that mean? Are these now functions outside of that spec? 😒
> > 
> > 
> > Technically, the ISO standard does not define any math functions beyond mixed integer-and-fixed-point multiplications and divisions, not even including `sqrt`.
> > Ideally, I would like to have all C23 math functions available for all numeric types. We could have a complete math library for fixed point type and with the proof of their usefulness, maybe fixed point types + its math library will be incorporated into the future C standards, and ours can be served as a reference implementation?
> 
> Can we make that more explicit on https://libc.llvm.org/math/stdfix.html?

I've updated stdfix's documentations.

File: libc/src/stdfix/CMakeLists.txt [hunk 3]
```diff
@@ -46,3 +46,29 @@ foreach(suffix IN ITEMS hr r lr hk k lk uhr ur ulr uhk uk ulk)
...
+
+add_entrypoint_object(
+  uhksqrtus
+  HDRS
+    uhksqrtus.h
+  SRCS
+    uhksqrtus.cpp
+  COMPILE_OPTIONS
+    -O3
+    -ffixed-point
```
Comment (michaelrj-google) [inline libc/src/stdfix/CMakeLists.txt:58]: if `LIBC_COMPILER_HAS_FIXED_POINT` is specified, then `-ffixed-point` is automatically added to all of the compile options. You don't have to additionally specify it here.

Comment (lntue) [reply libc/src/stdfix/CMakeLists.txt]: Removed all redundant `-ffixed-point` flags.

File: libc/docs/math/stdfix.rst [hunk 4]
```diff
@@ -93,6 +98,9 @@ roundfx
 Higher math functions
 =====================
 
+The following math functions are modeled after C99/C23 math functions for
+floating point types.
```
Comment (nickdesaulniers) [inline libc/docs/math/stdfix.rst:102]:
```diff
- floating point types.
+ floating point types, but are not part of the the ISO/IEC TR 18037:2008 spec.
```

**Review by nickdesaulniers — APPROVED** [review]

File: libc/docs/math/stdfix.rst (hunk 4 above)
Comment (lntue) [reply libc/docs/math/stdfix.rst]: Done.

File: libc/spec/llvm_libc_ext.td [hunk 5]
```diff
@@ -51,9 +51,29 @@ def LLVMLibcExt : StandardSpec<"llvm_libc_ext"> {
...
+          GuardedFunctionSpec<"uksqrtui", RetValSpec<UnsignedAccumType>, [ArgSpec<UnsignedIntType>], "LIBC_COMPILER_HAS_FIXED_POINT">,
+      ]
+  >;
+
   let Headers = [
-    Strings,
-    Sched,
     Assert,
+    Sched,
+    Stdfix,
```
Comment (nickdesaulniers) [inline libc/spec/llvm_libc_ext.td:76]: error: Variable not defined: 'Stdfix'
    Stdfix,

https://lab.llvm.org/buildbot/#/builders/163/builds/52530

Comment (lntue) [reply libc/spec/llvm_libc_ext.td:76]: https://github.com/llvm/llvm-project/pull/84259

**Comment by lntue** [comment]
> runtime build for armv6 baremetal target

Can you check if https://github.com/llvm/llvm-project/pull/84365 fixes the build issue?

**Comment by nickdesaulniers** [comment]
Orthogonally, once this is fixed, if you have guidance on how you configure the build of llvm libc for this target, we could look into adding such a build into postsubmit (even if we don't have that specific device, we can still test building for it, even until we have emulators in place to test with).

**Comment by zeroomega** [comment]
> Orthogonally, once this is fixed, if you have guidance on how you configure the build of llvm libc for this target, we could look into adding such a build into postsubmit (even if we don't have that specific device, we can still test building for it, even until we have emulators in place to test with).

For armv6m barebone target runtime setup, we use configurations at https://github.com/llvm/llvm-project/blob/2bd369b48dbf0bc3128becb7ef8f8a1b82514b87/clang/cmake/caches/Fuchsia-stage2.cmake#L303 . The runtime was built along side with Clang from source. You might need to change the flags a bit if you want to build the runtime stand alone. @petrhosek knows more about this than me.

Builder https://lab.llvm.org/buildbot/#/builders/98 is already a post submit builder and it should have sent out emails when it fails. Though this builder isn't specifically for barebone runtimes, we just happen to build it along side linux and fuchsia runtimes. Are you looking into adding a presubmit builder for armv6m barebone target?

**Pull request merged**